iroha/src/sumeragi.rs

```diff
@@ -123,209 +119,122 @@ impl Sumeragi {
-                "Block height should be 0 for genesis round. But it is: {}",
-                self.block_height
-            ))
-        } else {
-            self.validate_and_publish_created_block(
-                PendingBlock::new(transactions).chain_first_with_genesis_topology(genesis_topology),
-            )
-            .await
-        }
-    }
-
-    /// The leader of each round just uses the transactions they have at hand to create a block.
+    /// the leader of each round just uses the transactions they have at hand to create a block
     pub async fn round(&mut self, transactions: Vec<AcceptedTransaction>) -> Result<(), String> {
         if transactions.is_empty() {
             return Ok(());
         }
         if let Role::Leader = self.network_topology.role(&self.peer_id) {
+            let wsv = self.world_state_view.clone();
             let block = PendingBlock::new(transactions).chain(
                 self.block_height,
                 self.latest_block_hash,
                 self.number_of_view_changes,
                 self.invalidated_blocks_hashes.clone(),
             );
-            self.validate_and_publish_created_block(block).await
+            let block = block.validate(&*wsv.read().await, &self.permissions_validator);
+            log::info!(
+                "{:?} - Created a block with hash {}.",
+                self.network_topology.role(&self.peer_id),
+                block.hash(),
+            );
+            for event in Vec::<Event>::from(&block.clone()) {
+                self.events_sender.send(event).await;
+            }
+            if !self.network_topology.is_consensus_required() {
+                self.commit_block(block).await;
+                Ok(())
+            } else {
+                *self.voting_block.write().await = Some(VotingBlock::new(block.clone()));
+                let message = Message::BlockCreated(block.clone().sign(&self.key_pair)?.into());
+                let recipient_peers = self.network_topology.sorted_peers.clone();
+                let mut send_futures = Vec::new();
+                for peer in &recipient_peers {
+                    if self.peer_id != *peer {
+                        send_futures.push(message.clone().send_to(peer));
+                    }
+                }
+                let results = futures::future::join_all(send_futures).await;
+                results
+                    .iter()
+                    .filter(|result| result.is_err())
+                    .for_each(|error_result| {
+                        log::error!("Failed to send BlockCreated messages: {:?}", error_result)
+                    });
+                Ok(())
+            }
         } else {
-            self.forward_transactions_to_leader(&transactions).await?;
-            self.gossip_transactions(&transactions).await
-        }
-    }
-
-    /// Forwards transactions to the leader and waits for receipts.
-    pub async fn forward_transactions_to_leader(
```

Comment:
May be returning `forward_transactions_to_leader` and `gossip_transactions` would increase readability